pkg/liquidity-mining/test/DistributionScheduler.test.ts

```diff
@@ -0,0 +1,394 @@
+import { advanceToTimestamp, currentTimestamp, WEEK } from '@balancer-labs/v2-helpers/src/time';
+import { BigNumberish, maxUint } from '@balancer-labs/v2-helpers/src/numbers';
+
+const MAX_UINT224 = maxUint(224);
+const HEAD = 0;
+const NULL = 0;
+
+const roundDownTimestamp = (timestamp: BigNumberish): BigNumber => {
+  return BigNumber.from(timestamp).div(WEEK).mul(WEEK);
+};
+
+const roundUpTimestamp = (timestamp: BigNumberish): BigNumber => {
+  return roundDownTimestamp(BigNumber.from(timestamp).add(WEEK).sub(1));
+};
+
+type RewardNode = {
+  amount: BigNumber;
+  nextNodeKey: number;
+};
+
+describe('DistributionScheduler', () => {
+  let rewardToken: Token;
+  let rewardTokenDistributor: Contract;
+  let distributionScheduler: Contract;
+
+  let caller: SignerWithAddress;
+
+  before('setup signers', async () => {
+    [, caller] = await ethers.getSigners();
+  });
+
+  sharedBeforeEach('deploy DistributionScheduler', async () => {
+    rewardTokenDistributor = await deploy('MockRewardTokenDistributor');
+    distributionScheduler = await deploy('DistributionScheduler');
+  });
+
+  sharedBeforeEach('deploy Token', async () => {
+    rewardToken = await Token.create('REWARD');
+    await rewardToken.mint(caller);
+    await rewardToken.approve(distributionScheduler, MAX_UINT256, { from: caller });
+  });
+
+  async function getRewardNode(startTime: BigNumberish): Promise<RewardNode> {
+    return distributionScheduler.getRewardNode(rewardTokenDistributor.address, rewardToken.address, startTime);
+  }
+
+  async function getNextNodeKey(nodeKey: BigNumberish): Promise<BigNumber> {
+    return BigNumber.from((await getRewardNode(nodeKey)).nextNodeKey);
+  }
+
+  async function scheduleDistribution(amount: BigNumberish, timestamp: BigNumberish): Promise<ContractReceipt> {
+    const tx = await distributionScheduler
+      .connect(caller)
+      .scheduleDistribution(rewardTokenDistributor.address, rewardToken.address, amount, timestamp);
+    return tx.wait();
+  }
+
+  describe('scheduleDistribution', () => {
+    context('when providing zero tokens', () => {
+      it('reverts', async () => {
+        await expect(scheduleDistribution(0, 0)).to.be.revertedWith('Must provide non-zero number of tokens');
+      });
+    });
+
+    context('when providing more tokens than fit in a uint224', () => {
+      it('reverts', async () => {
+        await expect(scheduleDistribution(MAX_UINT224.add(1), 0)).to.be.revertedWith('Reward amount overflow');
+      });
+    });
+
+    context("when distribution starts at a timestamp which doesn't fit in a uint32", () => {
+      it('reverts', async () => {
+        await expect(scheduleDistribution(1, MAX_UINT32.add(1))).to.be.revertedWith('Reward timestamp overflow');
+      });
+    });
+
+    context('when reward token does not exist on gauge', () => {
+      it('reverts', async () => {
+        await expect(scheduleDistribution(1, 0)).to.be.revertedWith('Reward token does not exist on gauge');
```

Comment:
I'd consider changing these hardcoded 1's to e.g. `const amount = 42`. This is specially useful in later tests.